### DISCUSSION TITLE
Title trace languages

### DISCUSSION THREAD

**Discussion by 0xC0D3D34D** [unexpectedpanda/retool#261] [Q&A]
Tried the trace option to see where a title was being selected. At the compare languages point, how is the implied language order chosen? It appears to contain a lot of unnecessary duplicates

`REF0083: Compare languages
Implied language order: En, En, En, En, En, En, En, En, En, Ja, En, Es, Es, Es, Pt, Pt, Fr, Fr, Nl, De, De, It, Zh, Zh, Zh, Ko, Ru, Et, Pl, Lv, Lt, Da, No, Sv, Fi, Is, Hu, Cs, El, Mk, En, He, Tr, Hr, Sl, Ar, Bg, Ro, Sq, Sr, Id, En, En`

**Comment by unexpectedpanda** [comment]
It's based on your region order. The duplicates get condensed down before any actual processing happens, but that reporting is potentially misleading. I'll take a look when time presents.

**Reply by 0xC0D3D34D** [reply]
Changing region order doesn't seem to be having any effect on the implied language order

**Reply by unexpectedpanda** [reply]
Interesting. Have you also set a language order? It might be coming from the default region order, depending on the scenario.

(We might already be accelerating toward me getting your `user-config.yaml` file and any system YAML file to understand your setup, and finding out the title you're trying to trace in which particular DAT file)

**Reply by 0xC0D3D34D** [reply]
No particular game. I was just experimenting with the tracing to understand the inner workings. It occurred to me that with my language set to English and regions prioritized for Europe, some time there would be more than one match. For some reason there are games with multiple European versions containing English or English implied.

e.g.
```
Game1 (Europe)
Game1 (Europe) (En,Fr,De)
Game2 (Europe) (En,Fr,De)
Game2 (Europe) (En,Es,It)
```

I was curious how retool chooses between. Does setting a language stop the region order affecting the language list?

**Reply by unexpectedpanda** [reply]
Ah, so in my head there's a few different implied language orders going on. That's terminology I need to improve upon. After going through the code, this is how it works.

There are essentially two language orders: the user language order, and the default language order.

If the user chooses regions but not languages, the user language order is built out of the implied languages associated with each region they've added to their list. These implied languages are defined in `internal-config.json`. So if they only add a region order of USA > Europe > Italy, their user language order is English > Italian.

If the user chooses languages as well, then the user language order is instead built from those languages. So if they choose French, then the user language order becomes just French, regardless of the regions chosen.

In both circumstances the user can add as few regions or languages as they like, which can create a language order that's missing languages  -- a problem when trying to decide between titles. For example, a user might choose a region order of USA > Europe > Italy, and no languages, which gives them a user language order of English > Italian. That's fine when you have something like this:

```
Game 3 (Europe)
Game 3 (Europe) (It)
```

But then the situation can arise where Retool doesn't know what to choose when it sees something like this, that has none of the user languages:

```
Game 4 (Europe) (De,Da,Sv,No)
Game 4 (Europe) (Pt)
```

Or this, where there's no language in the order that can act as a tie breaker beyond English:

```
Game 5 (Europe) (En,De,Fr)
Game 5 (Europe) (En,De,Es,Fr)
```

So, the first stage language filter compares two titles against each other, going one by one through the languages in the user language order. If both titles have none of the languages in the user language order, or have all of the same languages, then Retool goes to the first fallback -- the default language order (built from the implied languages in the default region order, which is defined in `internal-config.json`). This is what's happening at REF0083, and this message could definitely be improved to help users understand better.

The default language order is used because Retool really can't get further data on what the user would prefer -- they haven't included the extra regions or languages to help it out. I've also made the assumption that if the user didn't care about adding the regions or languages themselves, then they probably don't care which variation of the game gets chosen in these scenarios. Arguably this could be a bit more nuanced -- if a user has defined a language order, then Retool could first attempt to fall back on the user language order defined by the regions before it gets to the default language order.

If both titles somehow contain none of the languages in the default language order, or both have the same languages (as can happen with clones with different names), then Retool progresses to the second fallback: choosing the title with the most languages.

If they both have the same number of languages, then the titles progress to the next stage of selection criteria.
